compiler/rustc_typeck/src/coherence/builtin.rs

```diff
@@ -543,17 +551,19 @@ pub fn coerce_unsized_info<'tcx>(tcx: TyCtxt<'tcx>, impl_did: DefId) -> CoerceUn
         let mut fulfill_cx = <dyn TraitEngine<'_>>::new(infcx.tcx);
 
         // Register an obligation for `A: Trait<B>`.
-        let cause = traits::ObligationCause::misc(span, impl_hir_id);
-        let predicate = predicate_for_trait_def(
-            tcx,
-            param_env,
-            cause,
-            trait_def_id,
-            0,
-            source,
-            &[target.into()],
-        );
-        fulfill_cx.register_predicate_obligation(&infcx, predicate);
+        if kind != CoerceUnsizedKind::JustMetadata {
```

Comment:
It's not immediately obvious why this predicate doesn't need to be registered, can you leave a comment explaining why?

Reply:
oh -- is it because of the `CoerceUnsized<JustMetadata<U>> for JustMetadata<T>`?

Reply:
Yes: only the coercion of `JustMetadata` gets `CoerceUnsizedKind::JustMetadata`, and applying the predicate would require that `<T as Pointee>::Metadata: CoerceUnsized<<U as Pointee>::Metadata>` (applying the `CoerceUnsized` requirement to the type-dependent field) which we don't want in this case, as we implement the coercion directly for `JustMetadata`.

Reply:
...actually no, you're right, it should be `A: Unsize<B>` here. The reason I didn't realize this initially is that I was returning the wrong `A`/`B` for `JustMetadata`, being the field tys instead of the generic tys. The fields aren't related, but we *should* check that the generic tys are correctly related.